File: examples/tito_example.py
"""Together In Together Out - Python example.

Run: python examples/tito_example.py
"""

from collections import OrderedDict, deque


class TitoQueue:
    """Minimal strict-order (head-of-line FIFO) TITO queue."""

    def __init__(self):
        self._order = deque()
        self._groups = {}
        self._group_of = {}

    def admit(self, group_id, members):
        members = list(members)
        if not members:
            raise ValueError("a group must contain at least one member")
        if group_id in self._groups:
            raise ValueError("duplicate group id: %r" % (group_id,))
        ready = OrderedDict()
        for member in members:
            if member in self._group_of or member in ready:
                raise ValueError("duplicate member: %r" % (member,))
            ready[member] = False
        for member in members:
            self._group_of[member] = group_id
        self._groups[group_id] = ready
        self._order.append(group_id)

    def mark_ready(self, member):
        group_id = self._group_of.get(member)
        if group_id is None:
            raise ValueError("unknown member: %r" % (member,))
        self._groups[group_id][member] = True

    def release(self):
        """Release the head group, or None while it is not fully ready."""
        if not self._order:
            return None
        group_id = self._order[0]
        ready = self._groups[group_id]
        if not all(ready.values()):
            return None
        self._order.popleft()
        del self._groups[group_id]
        for member in ready:
            del self._group_of[member]
        return group_id, list(ready)


def main():
    queue = TitoQueue()

    queue.admit("party-1", ["ann", "bob"])
    print("admit party-1: ann, bob")
    queue.admit("party-2", ["cy"])
    print("admit party-2: cy")

    for member in ("ann", "bob", "cy"):
        queue.mark_ready(member)
        print("ready %s" % member)
        released = queue.release()
        if released is None:
            print("release -> waiting")
        else:
            group_id, members = released
            print("release -> %s [%s]" % (group_id, " ".join(members)))


if __name__ == "__main__":
    main()

File: examples/tito_example.php
<?php
// Together In Together Out - PHP example.
//
// Run: php examples/tito_example.php

declare(strict_types=1);

/** Minimal strict-order (head-of-line FIFO) TITO queue. */
final class TitoQueue
{
    /** @var list<string> */
    private array $order = [];
    /** @var array<string, array<string, bool>> */
    private array $groups = [];
    /** @var array<string, string> */
    private array $groupOf = [];

    /** @param list<string> $members */
    public function admit(string $groupId, array $members): void
    {
        if ($members === []) {
            throw new InvalidArgumentException('a group must contain at least one member');
        }
        if (isset($this->groups[$groupId])) {
            throw new InvalidArgumentException("duplicate group id: $groupId");
        }
        $ready = [];
        foreach ($members as $member) {
            if (isset($this->groupOf[$member]) || isset($ready[$member])) {
                throw new InvalidArgumentException("duplicate member: $member");
            }
            $ready[$member] = false;
        }
        foreach ($members as $member) {
            $this->groupOf[$member] = $groupId;
        }
        $this->groups[$groupId] = $ready;
        $this->order[] = $groupId;
    }

    public function markReady(string $member): void
    {
        if (!isset($this->groupOf[$member])) {
            throw new InvalidArgumentException("unknown member: $member");
        }
        $this->groups[$this->groupOf[$member]][$member] = true;
    }

    /**
     * Release the head group, or null while it is not fully ready.
     *
     * @return array{0: string, 1: list<string>}|null
     */
    public function release(): ?array
    {
        if ($this->order === []) {
            return null;
        }
        $groupId = $this->order[0];
        $ready = $this->groups[$groupId];
        foreach ($ready as $isReady) {
            if (!$isReady) {
                return null;
            }
        }
        array_shift($this->order);
        unset($this->groups[$groupId]);
        $members = array_keys($ready);
        foreach ($members as $member) {
            unset($this->groupOf[$member]);
        }
        return [$groupId, $members];
    }
}

$queue = new TitoQueue();

$queue->admit('party-1', ['ann', 'bob']);
echo "admit party-1: ann, bob\n";
$queue->admit('party-2', ['cy']);
echo "admit party-2: cy\n";

foreach (['ann', 'bob', 'cy'] as $member) {
    $queue->markReady($member);
    echo "ready $member\n";
    $released = $queue->release();
    if ($released === null) {
        echo "release -> waiting\n";
    } else {
        [$groupId, $members] = $released;
        echo "release -> $groupId [" . implode(' ', $members) . "]\n";
    }
}
